### PR TITLE
Added Java maze

### PR DESCRIPTION
**Pull request by chrisjmendoza**
Added Java classes which will draw two mazes based on two different algorithms.

### CODE REVIEW

File: Java/Mazes/src/Cell.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    int getyCoor() {
+        return yCoor;
+    }
+
+
+
+    /**
+     * Checks that the number of available moves is greater than 0
+     *
+     * @return True is available populateAvailableDirections is greater than 0
```
Comment (dhoehna) [inline Java/Mazes/src/Cell.java]: Change this comment to indeicate what true means, not the condition under when the return value is true.  I can see by reading the code that, if the size of avalibleDirections is greater than 0 it should return true.

File: Java/Mazes/src/Cell.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     * Checks the cell for available populateAvailableDirections. If the Cell is along the border, it removes the edge as an available
+     * space
+     *
+     * @param height The height of the grid
+     * @param width  The width of the grid
+     */
+    void populateAvailableDirections(int height, int width) {
+        // Check upward movement
+        if (yCoor - 1 >= 0) {
+            availableDirections.add(1);
```
Comment (dhoehna) [inline Java/Mazes/src/Cell.java:66]: What do these magic numbers mean?

File: Java/Mazes/src/DrawMaze.java [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+    /**
+     * Construct a simple maze
+     *
+     * @param width  The width of the maze in "cells"
+     * @param height The height of the maze in "cells"
+     */
+    private DrawMaze(int width, int height) {
+        mazeWidth = width;
+        mazeHeight = height;
+        setPreferredSize(new Dimension(width * SIZE + 2 * BORDER + WALL,
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: Please add () to these equations to show calculation order.

File: Java/Mazes/src/DrawMaze.java [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    /**
+     * Add a vertical wall to the maze
+     *
+     * @param x   The horizontal offset for the wall
+     * @param y   The vertical offset for the wall
+     * @param len The length of the wall
+     * @return true, if the requested wall added to the set of walls
+     */
+    private boolean addVerticalWall(int x, int y, int len) {
+        if (x > mazeWidth || y + len > mazeHeight) throw new IllegalArgumentException("Wall exceeds maze boundary");
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: Please add {} to single line if blocks.

File: Java/Mazes/src/DrawMaze.java [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+    /**
+     * Add a vertical wall to the maze
+     *
+     * @param x   The horizontal offset for the wall
+     * @param y   The vertical offset for the wall
+     * @param len The length of the wall
+     * @return true, if the requested wall added to the set of walls
+     */
+    private boolean addVerticalWall(int x, int y, int len) {
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: Please no single letter variable names please.

File: Java/Mazes/src/DrawMaze.java [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        if (x > mazeWidth || y + len > mazeHeight) throw new IllegalArgumentException("Wall exceeds maze boundary");
+        boolean added = false;
+        for (int i = 0; i < len; i++) {
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: What are you iterating over?  I know it is the length of the wall, but, is it pixles, or something else?

File: Java/Mazes/src/DrawMaze.java [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+     *
+     * @return the set of walls
+     */
+    public Set<Wall> getWallSet() {
+        return walls;
+    }
+
+    void maze1() {
+        ArrayList<Point> source = new ArrayList<>();
+        ArrayList<Point> loose = new ArrayList<>();
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: Please move loose as close as possible to the first place it is being used.

Comment (dhoehna) [reply Java/Mazes/src/DrawMaze.java]: Also, what does loose stand for?

File: Java/Mazes/src/DrawMaze.java [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+
+                cellGrid.add(new Cell(i, j));
+            }
+        }
+
+        // Remove the wall on the opening
+        walls.remove(new Wall(0, 0, true));
+
+        // Create the stack to use for recursive pathfinding
+        Stack<Cell> stack = new Stack<>();
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java:287]: IF you can think of a better name for this stack than "Stack" you won't need to add this comment.

File: Java/Mazes/src/DrawMaze.java [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+        // Start at the last cell which will be the finish
+        Cell currentCell = cellGrid.get(cellGrid.size() - 1);
+
+        // In each cell, populate which populateAvailableDirections are available
```
Comment (dhoehna) [inline Java/Mazes/src/DrawMaze.java]: You don't need this comment the code says it all.

File: Java/Mazes/src/Wall.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    boolean horz; // Orientation
+
+    /**
+     * Create a wall object
+     * @param xcoord the x-coordinate for the wall
+     * @param ycoord the y-coordinate for the wall
+     * @param horizontal true, if the wall extends to the right of the point, false if it extends down
+     */
+    Wall(int xcoord, int ycoord, boolean horizontal) {
+        if(xcoord < 0 || ycoord < 0) {
```
Comment (dhoehna) [inline Java/Mazes/src/Wall.java:16]: NIT: You can change these into unsigned integers, that way you won't have to do this check.

File: Java/Mazes/src/Wall.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            return false;
+        }
+    }
+
+    /**
+     * Override hashCode for use in the map
+     * @return A hashcode for use in the map
+     */
+    @Override
+    public int hashCode() {
```
Comment (dhoehna) [inline Java/Mazes/src/Wall.java:44]: Do you actually use this function?

File: Java/Mazes/src/DrawMaze.java (hunk 6 above)
Comment (chrisjmendoza) [reply Java/Mazes/src/DrawMaze.java]: This was originally going to be used for error checking in one maze generating algorithm but then I ended up not calling it. It involved varying length walls and confirming that the length of wall was valid. I have removed it now and will reimplement if the times comes that another algorithm has use of it.

File: Java/Mazes/src/Cell.java (hunk 2 above)
Comment (chrisjmendoza) [reply Java/Mazes/src/Cell.java:66]: I added comments to clarify their meaning. I added a TODO comment for replacing these with enum later on.

File: Java/Mazes/src/Wall.java (hunk 11 above)
Comment (chrisjmendoza) [reply Java/Mazes/src/Wall.java:44]: If you remove the hashCode method, the first maze generating algorithm doesn't work. It just creates a grid.

**Comment by chrisjmendoza** [comment]
I will make changes and resubmit pull request

**Pull request closed without merging**